SpatialGDK/Source/SpatialGDK/Private/Utils/EntityPool.cpp

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "Utils/EntityPool.h"
+
+#include "TimerManager.h"
+
+#include "Interop/SpatialReceiver.h"
+#include "SpatialGDKSettings.h"
+
+DEFINE_LOG_CATEGORY(LogSpatialEntityPool);
+
+using namespace improbable;
+
+void UEntityPool::Init(USpatialNetDriver* InNetDriver, FTimerManager* InTimerManager)
+{
+	NetDriver = InNetDriver;
+	Receiver = InNetDriver->Receiver;
+	TimerManager = InTimerManager;
+
+	ReserveEntityIDs(GetDefault<USpatialGDKSettings>()->EntityPoolInitialReservationCount);
+}
+
+void UEntityPool::ReserveEntityIDs(int32 EntitiesToReserve)
+{
+	UE_LOG(LogSpatialEntityPool, Verbose, TEXT("Sending bulk entity ID Reservation Request"));
+
+	checkf(!bIsAwaitingResponse, TEXT("Trying to reserve Entity IDs while another reserve request is in flight"));
+
+	// Set up reserve IDs delegate
+	ReserveEntityIDsDelegate CacheEntityIDsDelegate;
+	CacheEntityIDsDelegate.BindLambda([EntitiesToReserve, this](Worker_ReserveEntityIdsResponseOp& Op)
+	{
+		if (Op.status_code != WORKER_STATUS_CODE_SUCCESS)
+		{
+			// UNR-630 - Temporary hack to avoid failure to reserve entities due to timeout on large maps
+			if (Op.status_code == WORKER_STATUS_CODE_TIMEOUT)
+			{
+				UE_LOG(LogSpatialEntityPool, Warning, TEXT("Failed to reserve entity IDs Reason: %s. Retrying..."), UTF8_TO_TCHAR(Op.message));
+				ReserveEntityIDs(EntitiesToReserve);
+			}
+			else
+			{
+				UE_LOG(LogSpatialEntityPool, Error, TEXT("Failed to reserve entity IDs Reason: %s."), UTF8_TO_TCHAR(Op.message));
+			}
+
+			return;
+		}
+
+		// Ensure we received the same number of reserved IDs as we requested
+		check(EntitiesToReserve == Op.number_of_entity_ids);
+
+		// Clean up any expired Entity ranges
+		ReservedEntityIDRanges = ReservedEntityIDRanges.FilterByPredicate([](const EntityRange& Element)
+		{
+			return !Element.bExpired;
+		});
+
+		EntityRange NewEntityRange = {};
```

Comment:
Initialise bExpired to false.

Reply:
does this not 0 initialise the struct and implicitly set bExpired to false?

Reply:
Oh, yes that's right. Ignore me.